File: 0002-add-two-numbers/0002-add-two-numbers.cpp

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* dummy = new ListNode(-1);
        ListNode* temp = dummy;

        ListNode* t1 = l1;
        ListNode* t2 = l2;

        int carry = 0;
        while (t1 || t2) {

            int a = t1 ? t1->val : 0;
            int b = t2 ? t2->val : 0;

            int sum = a + b + carry;
            temp->next = new ListNode(sum % 10);
            temp = temp->next;

            carry = sum / 10;
            t1 = t1 ? t1->next : NULL;
            t2 = t2 ? t2->next : NULL;
        }
        if (carry > 0) {
            temp->next = new ListNode(carry);
        }

        ListNode* ans = dummy->next;
        delete dummy;

        return ans;
    }

    // ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    //     ListNode* ans = new ListNode(-1);
    //     ListNode* it = ans;
    //     int carry = 0;
    //     while (l1 || l2 || carry) {
    //         int a = l1 ? l1->val : 0;
    //         int b = l2 ? l2->val : 0;
    //         int sum = a + b + carry;
    //         int digit = sum % 10;
    //         carry = sum / 10;
    //         it->next = new ListNode(digit);
    //         it = it->next;

    //         l1 = l1 ? l1->next : NULL;
    //         l2 = l2 ? l2->next : NULL;
    //         // or use this
    //         // if (l1 != nullptr) l1 = l1->next;
    //         // else l1 = nullptr;

    //         // if (l2 != nullptr) l2 = l2->next;
    //         // else l2 = nullptr;
    //     }
    //     ListNode* finalans = ans->next;
    //     delete ans;
    //     return finalans;
    // }
};

File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, list1, list2):
            dummy = ListNode()
            current = dummy
            carry = 0

            # Traverse both linked lists until all numbers and carry are processed
            while list1 is not None or list2 is not None or carry:
                sum = carry  # Start with the carry from the previous addition
                
                # Add data from list1 if available
                if list1 is not None:
                    sum += list1.val
                    list1 = list1.next  # Move to the next node in list1

                # Add data from list2 if available
                if list2 is not None:
                    sum += list2.val
                    list2 = list2.next  # Move to the next node in list2

                # Update carry (the part that will be carried to the next node)
                carry = sum // 10

                # Create a new node with the current sum's last digit
                new_node = ListNode(sum % 10)

                # Link the new node to the result list
                current.next = new_node
                current = current.next  # Move current pointer to the new node

            # Return the result linked list, skipping the dummy node
            return dummy.next
        